### ISSUE TITLE
ccache not found

### ISSUE THREAD

**Issue by dehlen** [facebook/react-native#46126]
### Description

When opening Xcode not from the command line (via xed f.e.) but directly from Finder/Dock and the ccache option is enabled the build will fail because ccache can not be found even though it is installed. The issue exists because the PATH most certainly won't include the ccache binary (f.e. when installed via homebrew, it is located at /opt/homebrew/bin/ccache but Xcodes PATH does not include it). When setting the project up for ccache, the existence of ccache is checked. However this runs as part of pod install in a terminal environment with a potentially different PATH env variable. To fix this I think the ccache path found at pod install time should be passed to the ccache-clang and ccache-clang++.sh scripts.

### Steps to reproduce

1. Enable the ccache flag in your Podfile 
```ruby
react_native_post_install(
  installer,
  react_native_path,
  :ccache_enabled => true
)
```
2. Install ccache
3. Run pod install
4. Open Xcode from Dock/Finder
5. Open the projects xcworkspace
6. Build the project

An error message will be visible stating:
[root_path]/node_modules/react-native/scripts/xcode/ccache-clang.sh: line 14: exec: ccache: not found
Command CompileC failed with a nonzero exit code

### React Native Version

0.74.2

### Affected Platforms

Build - MacOS

### Output of `npx react-native info`

```text
System:
  OS: macOS 14.6.1
  CPU: (8) arm64 Apple M1 Pro
  Memory: 84.19 MB / 32.00 GB
  Shell:
    version: "5.9"
    path: /bin/zsh
Binaries:
  Node:
    version: 20.11.0
    path: ~/.asdf/installs/nodejs/20.11.0/bin/node
  Yarn:
    version: 1.19.0
    path: ~/.asdf/shims/yarn
  npm:
    version: 10.2.4
    path: ~/.asdf/plugins/nodejs/shims/npm
  Watchman: Not Found
Managers:
  CocoaPods:
    version: 1.13.0
    path: /Users/David.vonKnobelsdorff/.asdf/shims/pod
SDKs:
  iOS SDK:
    Platforms:
      - DriverKit 23.5
      - iOS 17.5
      - macOS 14.5
      - tvOS 17.5
      - visionOS 1.2
      - watchOS 10.5
  Android SDK: Not Found
IDEs:
  Android Studio: 2024.1 AI-241.15989.150.2411.11948838
  Xcode:
    version: 15.4/15F31d
    path: /usr/bin/xcodebuild
Languages:
  Java:
    version: 17.0.7
    path: /usr/bin/javac
  Ruby:
    version: 3.2.4
    path: /Users/David.vonKnobelsdorff/.asdf/shims/ruby
npmPackages:
  "@react-native-community/cli": Not Found
  react: Not Found
  react-native: Not Found
  react-native-macos: Not Found
npmGlobalPackages:
  "*react-native*": Not Found
Android:
  hermesEnabled: true
  newArchEnabled: false
iOS:
  hermesEnabled: true
  newArchEnabled: false
```

### Stacktrace or Logs

```text
[root_path]/node_modules/react-native/scripts/xcode/ccache-clang.sh: line 14: exec: ccache: not found
Command CompileC failed with a nonzero exit code
```

### Reproducer

https://github.com/dehlen/rn-ccache-bug

**Comment by dehlen** [comment]
Checked with the latest version (0.75.2) and the issue persists.

**Comment by dehlen** [comment]
Yes that works since its run from the terminal where my path is set. Ccache is installed via homebrew and the installation directory is part of my terminal path. However macOS applications do not inherit the path and therefore ccache can not be found. I think what would mitigate this problem would be to add a user defined setting like CCHACHE_BINARY in the ruby script as part of the post_install hook where the path is found and printed out. Then update the shell script to use this variable instead of calling exec ccache.

I recognize that you can build the project without modifications however to build it in Xcode itself and not from the Terminal is broken for me when enabling cchache and i think this could potentially be fixed.

**Comment by dehlen** [comment]
I'll give it a shot. I won't be able to have a look at this before next week but will comment back with a PR next week.

**Comment by dehlen** [comment]
Sorry to comment again without having a PR ready for you. I just had a look at the problem. I think the .xcode.env (and .local) files would be the right place to input a new environment variable (CCACHE_BINARY). I can do this in scripts/cocoapods/utils.rb:create_xcode_env_if_missing. I would create a export for .xcode.env with `export CCACHE_BINARY=$(command -v ccache)` and for the local variant `echo 'export CCACHE_BINARY=#{ccache_path}' > #{file_path}.local` if ccache is available. I then would assume calling the ccache scripts via `with-environment.sh ccache_script` would work.

I would choose this behaviour for the following reasons:
1. If ccache is found in the PATH (running from the Terminal) the output of `command -v ccache` is used. If it is not in the PATH one has the ability to overwrite this in the .xcode.env.local. So no changes for existing users of this feature, unless you want to support this use case.
2. I think with-environment.sh and the .xcode environment files are exactly for this purpose
3. One could think about an error message like this:

```
CCACHE_BINARY="${CCACHE_BINARY:-$(command -v ccache || true)}"
if [ -z "$CCACHE_BINARY" ]; then
	echo "[Error] Could not find ccache. It looks like that the .xcode.env or .xcode.env.local " \
"files are misconfigured. Please check that they are exporting a valid CCACHE_BINARY " \
"variable, pointing to the ccache executable."
	exit 1
fi
```

However I am not quite sure why I can't make it run like that. Somehow when switching the CC, CXX, LD and LDPLUSPLUS variables to something like `/absolute/path/to/with-environment.sh "/absolute/path/to/clang/or/clang++-script.sh"` errors our for me although I added the CCACHE_BINARY export to my environment. Executing the with-environment.sh. I then tried to call with-environment.sh directly from the clang scripts without changing the way CC, CXX, LD and LDPLUSPLUS  are configured like this:

```
SCRIPT_DIR="$(dirname "$(readlink -f "$0")")"

REACT_NATIVE_CCACHE_CONFIGPATH=$SCRIPT_DIR/ccache.conf
# Provide our config file if none is already provided
export CCACHE_CONFIGPATH="${CCACHE_CONFIGPATH:-$REACT_NATIVE_CCACHE_CONFIGPATH}"

WITH_ENVIRONMENT="$SCRIPT_DIR/with-environment.sh"
. "$WITH_ENVIRONMENT"

CCACHE_BINARY="${CCACHE_BINARY:-$(command -v ccache || true)}"
if [ -z "$CCACHE_BINARY" ]; then
	echo "[Error] Could not find ccache. It looks like that the .xcode.env or .xcode.env.local " \
"files are misconfigured. Please check that they are exporting a valid CCACHE_BINARY " \
"variable, pointing to the ccache executable."
	exit 1
fi

exec "$CCACHE_BINARY" clang "$@"
```

However this does not work as well. The environment variables are not found f.e. PODS_ROOT is not found and therefore with-environment.sh does not find the xcode env files. I am willing to investigate a bit further but I feel like I am missing the kownledge regarding environment variables in Xcode and shell scripting to find a quick solution for now. I could pass the ccache path directly to the clang scripts however I am not sure whether this would be an acceptable solution?

**Comment by cipolleschi** [comment]
Hi @dehlen, thanks for investigating the solution.

The `with-environment.sh` should be effectively the best approach.
I don't know why it is not working for the `CC`, etc. settings... might it be because they expect a single script/executable with no parameters? It might be that Xcode uses that executable as starting point and passes extra parameters that do not work properly.

Having the error might help.

**Comment by dehlen** [comment]
Xcode passes parameters to ccache which is why I tried to call with-environment.sh from inside the ccache scripts. They pass all variables to ccache via $@. I will recreate my last approach and comment with the error messages.

**Comment by dehlen** [comment]
Another option might be to echo a warning if ccache can't be found and use clang or clang++ directly in that case. While this does not solve the issue at hand at least it would not result in a build issue. The user could then write some custom post_install hook or change the CC,LD environment variables to fix the problem locally. What do you think about this? I am not 100% sure which way would be preferred. Also it might be of course totally possible to find a better solution I haven't found yet. Open for any suggestions.

**Comment by cipolleschi** [comment]
The warning can be a good alternative, but I'm afraid that people might not see it.

**Comment by ArindamRayMukherjee** [comment]
Issue persists at `0.76.0` and what more, in a project that uses C++ turbomodules, it prevents from diagnosing upgrading issues (in my case 0.74.2 to 0.76.0).

After following the upgrade guide at `https://react-native-community.github.io/upgrade-helper/?from=0.74.2&to=0.76.0` , and doing pod install etc, the compilation process triggered by `react-native run-ios` (and xcode) fails with xcode error 65.
To diagnose the error, we need to go run the build process from XCode, but the reason above blocks me from doing that.

**Comment by builtbyproxy** [comment]
Throwing my two cents in, i've had this issue for a while now and found that clearing the contents of my `.xcode.env.local` fixed it. You can find more on the fix in this thread: https://github.com/facebook/react-native/issues/43285

**Comment by krystofwoldrich** [comment]
Hello everyone, I would love to be able to configure CCACHE_PATH from `.xcode.env.local`.

I'm having the same ccache not found error when build my app directly from Xcode. When using RN CLI the build works as expected.

**Comment by cipolleschi** [comment]
I understand that that's a problem, but I don't have time to work on this currently. 

It would be great if some of you can prepare a PR to fix the behavior, we can then iterate on that one.

**Comment by ste7en** [comment]
Hey everyone, I made a small contribution by fixing this error in #48257.
While it would be great to customise ccache's path in an .xcode.env(.local), I couldn't find a way to find and source this file in ccache scripts. Also I think this would be an improvement rather than a fix to the current issue.

**Comment by cipolleschi** [comment]
Thanks for doing that @ste7en!

**Comment by cebrakadabra** [comment]
If one needs this urgently, patching is your friend.
Move in changes from #48257 in your `node_modules`.

Step 1: If you don't have patch-package:

```
npm install patch-package -D 

```
add this to your package.json

```
"scripts": {
  "postinstall": "patch-package"
 }
```

Step 2:
test if it works for you & create patch file:

```
npx patch-package react-native
```

apply patch

```
npm install or yarn
```

**Comment by elkinjosetm** [comment]
I found a solution that I think is a little bit less invasive and does not requires adding the CCACHE_BINARY path to the xcode project, but instead it pulls it all the time from the `.env` file, it's basically a combination of the solutions explained here.

1. Add CCACHE_BINARY to .xcode.env and .xcode.env.local
```
export CCACHE_BINARY=$(command -v ccache)
```
This allows you to customize the path on the `.local` file as you wish

2. Update `node_modules/react-native/scripts/xcode/ccache-clang.sh` by replacing the line `exec ccache clang "$@"` with `exec $CCACHE_BINARY clang "$@"`

3. Update `node_modules/react-native/scripts/xcode/ccache-clang++.sh` by replacing the line `exec ccache clang++ "$@"` with `exec $CCACHE_BINARY clang++ "$@"` 


Ultimately I patched react-native while an official solution is provided
